src/java/org/apache/cassandra/service/accord/AccordConfigurationService.java

```diff
@@ -614,28 +636,32 @@ public static EpochSnapshot notStarted(long epoch)
     }
 
     @VisibleForTesting
-    public synchronized EpochSnapshot getEpochSnapshot(long epoch)
+    public EpochSnapshot getEpochSnapshot(long epoch)
     {
         if (epoch < epochs.minEpoch() || epoch > epochs.maxEpoch())
             return null;
 
-        return new EpochSnapshot(getOrCreateEpochState(epoch));
+        // lock for read visibility to syncStatus
+        synchronized (this)
+        {
+            return new EpochSnapshot(getOrCreateEpochState(epoch));
```

Comment:
do we want to hold both locks at once? can't we get the `EpochState` and then build `EpochSnapshot`? What visibility are we looking for here, isn't it enough to piggy-back on the visibility semantics of the `epoch` monitor?

Reply:
called `getOrCreateEpochState` outside the lock

> What visibility are we looking for her

Most call paths of `org.apache.cassandra.service.accord.AccordConfigurationService.EpochState#syncStatus` already hold the accord lock, so need this here else it can be stale (unless we make it volatile)

Reply:
Technically it has visibility due to the semantics of the `getOrCreateEpochState` - monitor entry has the semantics of a global read barrier. But this isn't important to optimise, and it's fine to make volatile.